test/remote/too_many_totp_codes.js

```diff
@@ -0,0 +1,96 @@
+/* Any copyright is dedicated to the Public Domain.
+ * http://creativecommons.org/publicdomain/zero/1.0/ */
+
+const test = require('tap').test
+const TestServer = require('../test_server')
+const Promise = require('bluebird')
+const restify = Promise.promisifyAll(require('restify'))
+const mcHelper = require('../memcache-helper')
+
+const TEST_IP = '192.0.2.1'
+const TEST_EMAIL = 'one@two.three'
+
+const config = {
+  listen: {
+    port: 7000
+  }
+}
+
+process.env.TOTP_CODE_RULE_MAX = 2
+process.env.TOTP_CODE_RULE_PERIOD_MS = 1000
+process.env.TOTP_CODE_RULE_LIMIT_INTERVAL_MS = 1000
+
+const action = 'verifyTotpCode'
+
+const testServer = new TestServer(config)
+
+const client = restify.createJsonClient({
+  url: 'http://127.0.0.1:' + config.listen.port
+})
+
+Promise.promisifyAll(client, {multiArgs: true})
+
+test('startup', (t) => {
+  testServer.start((err) => {
+    t.type(testServer.server, 'object', 'test server was started')
+    t.notOk(err, 'no errors were returned')
+    t.end()
+  })
+})
+
+test('clear everything', (t) => {
+  mcHelper.clearEverything((err) => {
+    t.notOk(err, 'no errors were returned')
+    t.end()
+  })
+})
+
+test('/checkCustomRule `' + action + '` by email', (t) => {
+  // Send requests until throttled
+  return client.postAsync('/checkCustomRule', {ip: TEST_IP, configName: 'totpCode', action: action, email: TEST_EMAIL})
+    .spread((req, res, obj) => {
+      t.equal(res.statusCode, 200, 'returns a 200')
+      t.equal(obj.block, false, 'not rate limited')
+      return client.postAsync('/checkCustomRule', {ip: TEST_IP, configName: 'totpCode', action: action, email: TEST_EMAIL})
+    })
+    .spread((req, res, obj) => {
+      t.equal(res.statusCode, 200, 'returns a 200')
+      t.equal(obj.block, false, 'not rate limited')
+      return client.postAsync('/checkCustomRule', {ip: TEST_IP, configName: 'totpCode', action: action, email: TEST_EMAIL})
+    })
+    .spread((req, res, obj) => {
+      t.equal(res.statusCode, 200, 'returns a 200')
+      t.equal(obj.block, true, 'rate limited')
+      t.equal(obj.retryAfter, 1, 'rate limit retry amount')
+
+      // Delay ~1s for rate limit to go away
+      return Promise.delay(1010)
+    })
+
+    // Reissue requests to verify that throttling is disabled
+    .then(() => {
+      return client.postAsync('/checkCustomRule', {ip: TEST_IP, configName: 'totpCode', action: action, email: TEST_EMAIL})
+    })
+    .spread((req, res, obj) => {
+      t.equal(res.statusCode, 200, 'returns a 200')
+      t.equal(obj.block, false, 'not rate limited')
+      t.end()
+    })
+    .catch((err) => {
+      t.fail(err)
+      t.end()
+    })
+})
```

Comment:
I think it'd be worth adding another test here, to ensure that the rate-limiting is independent of other rate-limiting actions.  For example, that a user can still attempt `accountLogin` actions after be rate-limited for totp verification, and conversely that a user rate-limited for `accountLogin` actions can still do totp verification attempts.